winch/codegen/src/isa/x64/masm.rs

```diff
@@ -418,6 +418,60 @@ impl Masm for MacroAssembler {
     fn jmp(&mut self, target: MachLabel) {
         self.asm.jmp(target);
     }
+
+    fn popcnt(&mut self, context: &mut CodeGenContext, size: OperandSize) {
+        let src = context.pop_to_reg(self, None, size);
+        if self.flags.has_popcnt() {
+            self.asm.popcnt(src, size);
+            context.stack.push(Val::reg(src));
+        } else {
+            let tmp = context.any_gpr(self);
+            let dst = src;
+            let (masks, shift_amt) = match size {
+                OperandSize::S64 => (
+                    [
+                        0x5555555555555555, // m1
+                        0x3333333333333333, // m2
+                        0x0f0f0f0f0f0f0f0f, // m4
+                        0x0101010101010101, // h01
+                    ],
+                    56u8,
+                ),
+                // 32-bit popcount is the same, except the masks are half as
+                // wide and we shift by 24 at the end rather than 56
+                OperandSize::S32 => (
+                    [0x55555555i64, 0x33333333i64, 0x0f0f0f0fi64, 0x01010101i64],
+                    24u8,
+                ),
+            };
```

Comment:
This is perfectly reasonable as-is but it keeps bothering me that the constant masks are duplicated in this way. One alternative might be:

```suggestion
            let masks = [
                0x5555555555555555, // m1
                0x3333333333333333, // m2
                0x0f0f0f0f0f0f0f0f, // m4
                0x0101010101010101, // h01
            ];
            let (mask, shift_amt) = match size {
                OperandSize::S64 => (u64::MAX as i64, 56u8),
                OperandSize::S32 => (u32::MAX as i64, 24u8),
            };
```

Then use e.g. `masks[0] & mask`. (Maybe rename `mask` though.)

Another approach is to generate the constants using bit-twiddling tricks. I don't think this is a _good_ idea but I spent enough time figuring out how it would work that I'm going to write it down anyway. You can divide `u64::MAX` or `u32::MAX` by certain constants to get these repeating patterns: specifically, `[0x3, 0x5, 0x11, 0xff]` to produce `[0x55..., 0x33..., 0x0f..., 0x01...]` respectively.

Reply:
Yup, I don't feel strongly about it either, so I'm good with leaving it as is. :+1: